ratelimit/src/lib.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+extern crate aio_limited;
+#[macro_use]
+extern crate futures;
+extern crate libp2p_swarm as swarm;
+#[macro_use]
+extern crate log;
+extern crate tokio;
+extern crate tokio_io;
+
+use aio_limited::{Limited, Limiter};
+use std::io;
+use swarm::{Multiaddr, Transport};
+use tokio::executor::Executor;
+use tokio::prelude::*;
+use tokio_io::io::{ReadHalf, WriteHalf};
+
+#[derive(Clone)]
+pub struct RateLimited<T> {
+    value: T,
+    rlimiter: Limiter,
+    wlimiter: Limiter,
+}
+
+impl<T> RateLimited<T> {
+    pub fn new<E: Executor>(
+        e: &mut E,
+        value: T,
+        max_read: usize,
+        max_write: usize,
+    ) -> io::Result<RateLimited<T>> {
+        Ok(RateLimited {
+            value,
+            rlimiter: Limiter::new(e, max_read).map_err(|e| {
+                error!("failed to create read limiter: {}", e);
```

Comment:
These should have a target I think, like `error!(target: "libp2p-ratelimit", "failed to create ...")`

I'm also conflicted about whether we want to use `error!` or a more discrete thing like `debug!`, since we're returning the error anyway.